File: diferencas_divididas_newthon.cpp
#include <bits/stdc++.h>

using namespace std;

#ifndef ONLINE_JUDGE
#include "debug.h"
#else
#define dbg(x...)
#endif

/* stop freaking out pls */

vector<long double> dif_div(vector<long double> &x, vector<long double> &y) {
    int n = y.size();
    vector<long double> c = {y[0]};
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            long double a = (y[j + 1] - y[j]) / (x[j + i + 1] - x[j]);
            y[j] = a;
        }
        c.emplace_back(y[0]);
    }
    return c;
}

long double poly(long double t, vector<long double>&x, vector<long double>&coef) {
    int n = x.size();
    long double ans = coef[0];
    for (int i = 0; i < n - 1; i++) {
        long double p = coef[i + 1];
        for (int j = 0; j <= i; j++) {
            p *= (t - x[j]);
        }
        ans += p;
    }
    return ans;
}

int main(){
    int n;
    cin >> n;
    vector<long double> x(n), y(n);
    for (auto &u : x) cin >> u;
    for (auto &u : y) cin >> u;
    vector<long double> coef = dif_div(x, y);
    cout << setprecision(15) << fixed;
    for (auto u : coef) cout << u << ',';
    cout << '\n';
    for (;;) {
        long double t;
        cin >> t;
        cout << poly(t, x, coef) << '\n';
    }
}

File: diferencas_divididas_newthon.py
import numpy as np
def dif_div(x, y):
    c = [y[0]]
    for i in range(len(y) - 1):
        for j in range(n - i - 1):
            y[j] = (y[j + 1] - y[j]) / (x[j + i + 1] - x[j])
        c.append(y[0])
    return c

def build_poly(t, x, coef):
    def poly(t):
        n = len(x)
        ans = coef[0]
        for i in range(n - 1):
            p = coef[i + 1]
            for j in range(i + 1):
                p *= (t - x[j]);
            ans += p;
        return ans
    return poly

def f(x):
    return np.cos(x)**3 + 2 * (np.cos(x)**2) + 1

n = int(input())
x = []
y = []
for i in range(n):
    t = float(input())
    x.append(t)
    y.append(f(t))

coef = dif_div(x, y)
for i in coef:
    print("%.15f" % i, end=",")
print()





